### PR TITLE
Add automatic differentiation for accelerated functions

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Functions that have been accelerated can be are compatible with JAX. This means that JAX can also auto-differentiate them. So, let's leverage JAX to auto-differentiate them.

**Description of the Change:** 

* Add a GradContext to determine when we are inside a grad statement
* If we are inside a GradContext then run a different path that will lead to defining the forward and reverse passes for accelerate.

**Benefits:** Better UI

**Possible Drawbacks:** Callback code is becoming harder to reason about with all the special cases, accelerate, pure_callback, with/without gradient.

**Related GitHub Issues:**

[sc-60775]

### CODE REVIEW

File: frontend/test/pytest/test_callback.py [hunk 1]
```diff
@@ -1390,6 +1390,39 @@ def ground_truth(x):
...
+    @jax.jacobian
+    def ground_truth(x):
+        return jax.scipy.linalg.expm(x)
+
+    if order == "bad":
+        obs = hypothesis(arg)
+        exp = ground_truth(arg)
+    else:
+        exp = ground_truth(arg)
+        obs = hypothesis(arg)
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_callback.py:1411]: This actually succeeded!

File: doc/changelog.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 <h3>New features</h3>
 
+* Automatic differentiation of `catalyst.accerate`-ed functions.
+  [(#920)](https://github.com/PennyLaneAI/catalyst/pull/920)
```
Comment (josh146) [inline doc/changelog.md:6]: Hey Erick, just a friendly reminder that there is a changelog entry pre-written in the epic (would appreciate also adding this example as a test to ensure it works, if it is not added already!)

File: frontend/catalyst/api_extensions/callbacks.py
Comment (josh146) [inline frontend/catalyst/api_extensions/callbacks.py:1]: @erick-xanadu maybe I missed it in this file, but has the docstring for `accelerate` been updated? previously it explicitly says that differentiation is not supported

Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:1]: I'll update it! Thanks for catching this :)

File: doc/changelog.md (hunk 2 above)
Comment (erick-xanadu) [reply doc/changelog.md]: Added example from the epic and used this changelog entry. :)

File: frontend/catalyst/api_extensions/callbacks.py
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:1]: Updated!

Comment (josh146) [reply frontend/catalyst/api_extensions/callbacks.py:1]: thanks!

File: doc/changelog.md [hunk 3]
```diff
@@ -2,6 +2,26 @@
 
 <h3>New features</h3>
 
+* JAX-compatible functions which run on classical accelerators such as GPUs via catalyst.accelerate now support autodifferentiation.
```
Comment (josh146) [inline doc/changelog.md:5]:
```suggestion
* JAX-compatible functions which run on classical accelerators such as GPUs via `catalyst.accelerate` now support autodifferentiation.
```

File: frontend/catalyst/api_extensions/callbacks.py [hunk 4]
```diff
@@ -327,21 +350,36 @@ def total(context, *args, **kwargs):
...
+
+            @with_custom_grad.fwd
+            @accelerate(dev=dev)
+            def vjp_wrapper(context, *args, **kwargs):
+                return jax.vjp(jitted_fn, context, *args, **kwargs)
+
+            @with_custom_grad.bwd
+            @accelerate(dev=dev)
+            def reverse(vjp_func, dy):
+                return vjp_func(dy)
```
Comment (dime10) [inline frontend/catalyst/api_extensions/callbacks.py:380]: It probably is, but I'm curious if the context is required:
I'm guessing you want to avoid unnecessary tracing of the function if it's not needed. Is there a chance we could define the attributes and the functions would just be traced on demand?

File: frontend/catalyst/tracing/contexts.py [hunk 5]
```diff
@@ -41,6 +41,79 @@
 logger.addHandler(logging.NullHandler())
 
 
+class GradContext:
+    """This class tells us the level of nestedness of grad."""
```
Comment (dime10) [inline frontend/catalyst/tracing/contexts.py:45]: Neat! Also thanks for the detailed comment

File: frontend/catalyst/tracing/contexts.py [hunk 6]
```diff
@@ -41,6 +41,79 @@
...
+    #
+    # The final question to be asked here is whether we need a new stack for this
+    # context or whether we should reuse the old one.
+    # I think we should keep it simple and this is simpler.
+
+    _grad_stack: int = 0
+    # This message will be used in an assertion because it is not expected
+    # to be a user facing error ever.
+    _error_msg: str = """This is an impossible state. One cannot derive to
+derive to a negative order / integrate"""
```
Comment (dime10) [inline frontend/catalyst/tracing/contexts.py:83]: Minor: Is there a reason for this message to be a class variable? (locality)

File: frontend/test/pytest/test_contexts.py
Comment (dime10) [inline frontend/test/pytest/test_contexts.py:1]: Thanks for adding unit tests!

**Review by dime10 — COMMENTED** [review]
Nice work 💯

File: frontend/catalyst/api_extensions/callbacks.py [hunk 7]
```diff
@@ -54,9 +58,10 @@ def accelerate(func=None, *, dev=None):
 
     .. note::
 
-        ``catalyst.accelerate`` doses not currently support
-        differentiation, and cannot be used inside functions that
-        :func:`catalyst.grad` is applied to.
+        ``catalyst.accelerate`` currently supports
+        differentiation, and can be used inside
+        :func:`catalyst.grad` and :func:`catalyst.jacobian` without one having
+        to define its gradient.
```
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/api_extensions/callbacks.py:64]: Yep I would just remove this note, while extending the doc string example to show differentiation. I can do that quickly.

[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/api_extensions/callbacks.py:64]: Edit: I see @erick-xanadu has already done the latter 🎉 so if we just remove this note that is good

File: frontend/catalyst/tracing/contexts.py (hunk 6 above)
Comment (erick-xanadu) [reply frontend/catalyst/tracing/contexts.py:83]: Ah, originally I had the assertion in two functions here.

File: frontend/catalyst/api_extensions/callbacks.py (hunk 4 above)
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:380]: >  Is there a chance we could define the attributes and the functions would just be traced on demand?

Potentially, one of the reasons the context check is required here is because accelerate now recurses. So if we are taking the derivative of function foo, without a check, we would also be taking the derivative of the derivative of foo.

So, a check is needed at some point, whether it is here or inside `with_custom_grad.__call__`. To me, it makes more sense for it to be here. If you want I can dedicate a few cycles to see if I can sink it.

File: frontend/catalyst/api_extensions/callbacks.py (hunk 7 above)
Comment (erick-xanadu) [inline frontend/catalyst/api_extensions/callbacks.py:64]:
```suggestion
```

File: frontend/catalyst/api_extensions/callbacks.py [hunk 8]
```diff
@@ -54,9 +58,10 @@ def accelerate(func=None, *, dev=None):
 
     .. note::
```
Comment (erick-xanadu) [inline frontend/catalyst/api_extensions/callbacks.py:60]:
```suggestion
```

File: frontend/catalyst/api_extensions/callbacks.py (hunk 7 above)
Comment (erick-xanadu) [inline frontend/catalyst/api_extensions/callbacks.py:64]:
```suggestion
```

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py]: Changed! https://github.com/PennyLaneAI/catalyst/pull/920/commits/5215636342d703fee811d8c765cabdbb46c8c9e5

File: frontend/catalyst/api_extensions/callbacks.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/callbacks.py:374]: > we would also be taking the derivative of the derivative of foo

Is that a bad thing? 🤔  I think that's how jax supports higher level derivates no?

> So, a check is needed at some point, whether it is here or inside with_custom_grad.__call__. To me, it makes more sense for it to be here. If you want I can dedicate a few cycles to see if I can sink it.

That's probably fine, I was just wondering whether we could always assign the fwd/bwd and if they would only be called upon demand.

Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:374]: > Is that a bad thing? 🤔 I think that's how jax supports higher level derivates no?

That's infinite derivatives if there is no check.

File: frontend/catalyst/api_extensions/callbacks.py [hunk 9]
```diff
@@ -327,21 +350,36 @@ def total(context, *args, **kwargs):
...
     # wraps total which wraps user
-    @functools.wraps(total)
+    @functools.wraps(total, assigned=assigned)
     def back_to_user(*args, **kwargs):
         absextra, absargs, abskwargs = tree_map(shaped_abstractify, (context, args, kwargs))
         try:
             # Find the shape of the return value
             with transient_jax_config({"jax_dynamic_shapes": False}):
-                _, returnshape = jax.make_jaxpr(total, return_shape=True)(
+                _, returnshape = jax.make_jaxpr(jitted_fn, return_shape=True)(
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:353]: Nothing, we just need the returnshape. I think jitted_fn makes more sense, but it is equivalent.

File: frontend/catalyst/api_extensions/callbacks.py [hunk 10]
```diff
@@ -251,19 +266,27 @@ def vjp(x, dy) -> (jax.ShapeDtypeStruct((2,), jnp.float64),):
...
 
 
 ## IMPL ##
-class AnnotatedFunction:
+class AnnotatedFunction(ABC):
+    """Defining an interface"""
+
+    def getResultTypes(self):
+        """Get result type of function"""
+        ...  # pragma: nocover
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py]: https://github.com/PennyLaneAI/catalyst/pull/920/commits/af8d51f668c095223a2eb344159e226c626427d2

File: frontend/catalyst/tracing/contexts.py (hunk 6 above)
Comment (erick-xanadu) [reply frontend/catalyst/tracing/contexts.py]: https://github.com/PennyLaneAI/catalyst/pull/920/commits/1e543538dcfb28dde111faed790cd5ab58790fe5

File: frontend/catalyst/api_extensions/callbacks.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/callbacks.py:375]: Oh I see, right. You'd have to use jax.jit instead of accelerate to avoid the infinite recursion I suppose. Thanks!

File: frontend/catalyst/api_extensions/callbacks.py [hunk 11]
```diff
@@ -327,21 +350,36 @@ def total(context, *args, **kwargs):
         jitted_fn = jax.jit(total)
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:355]: https://github.com/PennyLaneAI/catalyst/pull/920/commits/05957b7fc184d0ea85ebf76ab8c1a971290cc52f

File: frontend/catalyst/api_extensions/callbacks.py [hunk 12]
```diff
@@ -46,6 +47,9 @@
 )
 from catalyst.utils.types import convert_pytype_to_shaped_array
 
+# THIS IS NEEDED TO AVOID AUTOGRAPH CONVERSION
+assigned = list(filter(lambda x: x != "__module__", functools.WRAPPER_ASSIGNMENTS))
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/callbacks.py]: perfect 👍

File: frontend/catalyst/tracing/contexts.py (hunk 6 above)
Comment (erick-xanadu) [reply frontend/catalyst/tracing/contexts.py]: Ah, the other reason was to avoid having to update the test if the error message changes. See this change I had to add once the _error_msg was moved to the function:

https://github.com/PennyLaneAI/catalyst/pull/920/commits/792b0f1bb1368f7af8caf033a32079fb2a0a9ac5

**Review by dime10 — APPROVED** [review]
